### PR TITLE
New Relic Plugin Base Improvements

### PR DESCRIPTION
**Pull request by iamEAP**
## What/Why

Before hacking on what I outlined in [#824](https://github.com/spotify/backstage/issues/824#issuecomment-705446519), I wanted to address a couple of issues in the existing plugin:

- The REST API Key is exposed on the client-side, which is not great...
- If a key was not properly configured, a generic JavaScript error was shown.

This PR moves API communication to the proxy, and handles errors in a slightly friendlier way.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

**Comment by dtuite** [comment]
@iamEAP Thanks for this. Coincidentally, I was trying to get newrelic working today and I was pretty confused by the API Key setup.

I submitted a PR about the `NEW_RELIC_REST_API_KEY` the other day. https://github.com/spotify/backstage/pull/2875/files This should work with this change right?

**Comment by iamEAP** [comment]
@dtuite I'm not super familiar with helm, but if your configs mean that the Backstage backend is spun up with the `NEW_RELIC_REST_API_KEY` variable set, then these changes should be compatible.

Perhaps some of the initial trouble you ran into was because, prior to this change, it wasn't explicitly pulling the API Key in from the env?

**Comment by dtuite** [comment]
@iamEAP Exactly. I would end up with the frontend trying to use `NEW_RELIC_REST_API_KEY` as the API key rather than the env var which was set at that location. 😄

**Review by dtuite — APPROVED** [review]
LGTM and would have helped me out earlier today.

**Review by Rugvip — COMMENTED** [review]
👌, small nit

File: plugins/newrelic/src/api/index.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  application_instances: Array<any>;
+  servers: Array<any>;
+  application_hosts: Array<any>;
+};
+
+export type NewRelicApplications = {
+  applications: NewRelicApplication[];
+};
+
+export const newRelicApiRef = createApiRef<NewRelicApi>({
```
[the remark this answers is not in the document]
Comment (iamEAP) [reply plugins/newrelic/src/api/index.ts:56]: Ah, nice.  Good opportunity for me to read through the docs on this subsystem.  Make sense!

I made `NewRelicApi` an interface that `NewRelicClient` implements and used the interface in the `createApiRef` typing.

**Review by Rugvip — APPROVED** [review]
👍

File: plugins/newrelic/src/api/index.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  async getApplications(): Promise<NewRelicApplications> {
+    const url = await this.getApiUrl('apm', 'applications.json');
+    const response = await fetch(url);
+    const responseJson = await response.json();
+
+    if (response.status !== 200) {
+      throw new Error(responseJson?.error?.title || response.statusText);
+    }
```
Comment (freben) [inline plugins/newrelic/src/api/index.ts:91]: bit of a nit, but `.json()` explodes if it wasn't a 2xx response with data i think, so you may want to do this check right after the fetch instead (or twice, in different ways, whichever is best)

Comment (freben) [reply plugins/newrelic/src/api/index.ts:91]: Then again maybe it's ok the way you have it now, if it's known that the api always returns a json body no matter what, and it's fine to explode on the very unexpected corner case

Comment (iamEAP) [reply plugins/newrelic/src/api/index.ts:91]: That's a good call-out; if we're going to be friendlier with error messages, we may as well say so when New Relic is hard down too.

Comment (iamEAP) [reply plugins/newrelic/src/api/index.ts:91]: Probably good to add a little extra context in any event.

**When `api.newrelic.com` is blackhole'd**:
Screenshot 2020-10-15 at 12 28 32

**When the key is invalid**:
Screenshot 2020-10-15 at 12 36 00

File: plugins/newrelic/README.md [hunk 3]
```diff
@@ -7,16 +7,38 @@ Website: [https://newrelic.com](https://newrelic.com)
...
+you've set the `NEW_RELIC_REST_API_KEY` environment variable before starting
+the backend.
+
+While working locally, you may wish to hard-code your API key in your
+`app-config.local.yaml` like this:
+
+```yaml
+# app-config.local.yaml
+proxy:
+  'newrelic/apm/api':
```
Comment (freben) [inline plugins/newrelic/README.md:32]: leading slash missing

**Review by freben — APPROVED** [review]
aaaalllllmost there!

File: plugins/newrelic/README.md (hunk 3 above)
Comment (iamEAP) [reply plugins/newrelic/README.md]: 🦅  👀

**Review by freben — APPROVED** [review]

**Pull request merged**